packages/synapse-react-client/src/components/SchemaDrivenAnnotationEditor/field/AdditionalPropertiesSchemaField.tsx

```diff
@@ -0,0 +1,282 @@
+import {
+  FieldProps,
+  FormContextType,
+  RJSFSchema,
+  StrictRJSFSchema,
+} from '@rjsf/utils'
+import { isEqual } from 'lodash-es'
+import React, { useEffect, useState } from 'react'
+import FullWidthAlert from '../../FullWidthAlert/FullWidthAlert'
+import { convertToArray } from '../AnnotationEditorUtils'
+import { Grid, InputLabel } from '@mui/material'
+import { AdditionalPropertyContextProvider } from '../template/AdditionalPropertyContext'
+
+// Matches ####-##-##T##:##:##.###Z, e.g. 1970-01-01T12:00:000Z
+const ISO_TIMESTAMP_REGEX = /\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z/
+
+// Types that correspond to the different input fields that the annotation editor supports
+export enum PropertyType {
+  STRING = 'String',
+  INTEGER = 'Integer',
+  FLOAT = 'Float',
+  BOOLEAN = 'Boolean',
+  DATETIME = 'Datetime',
+}
+
+export function guessPropertyType(list: Array<unknown>): PropertyType {
+  if (list.length === 0) {
+    // The field was just added, so default to string
+    return PropertyType.STRING
+  } else if (
+    list.every(
+      item => typeof item === 'number' || item === 'NaN', // "NaN" is technically a float value
+    )
+  ) {
+    if (list.every(item => Number.isInteger(item))) {
+      return PropertyType.INTEGER
+    } else {
+      return PropertyType.FLOAT
+    }
+  } else if (list.every(item => typeof item === 'boolean')) {
+    return PropertyType.BOOLEAN
+  } else if (
+    list.every(item => typeof item === 'string') &&
+    (list as string[]).every((item: string) => !!ISO_TIMESTAMP_REGEX.exec(item))
+  ) {
+    return PropertyType.DATETIME
+  }
+  // otherwise, default type is 'string'
+  return PropertyType.STRING
+}
+
+export function transformDataFromPropertyType(
```

Comment:
Changed fallback/null cases from `undefined` to 0 (for integer), current date (for datetime) to reduce the likelihood of an empty property being dropped while edited. This wouldn't result in lost data but could be frustrating.